core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala

```diff
@@ -0,0 +1,207 @@
+    inWriteLock(lock) {
+      if (epoch >= 0 && epoch > latestEpoch && offset >= latestOffset) {
+        info(s"Updated PartitionLeaderEpoch ${epochChangeMsg(epoch, offset)}. Cache now contains ${epochs.size} entries.")
+        epochs += EpochEntry(epoch, offset)
+        flush()
+      }
+      maybeWarn(epoch, offset)
+    }
+  }
+
+  /**
+    * Returns the current Leader Epoch
+    *
+    * @return
+    */
+  override def latestEpoch(): Int = {
+    inReadLock(lock) {
+      if (epochs.isEmpty) UNDEFINED_EPOCH else epochs.last.epoch
+    }
+  }
+
+  /**
+    * Returns the End Offset for a requested Leader Epoch.
+    *
+    * This is defined as the start offset of the first Leader Epoch larger than the
+    * Leader Epoch requested, or else the Log End Offset if the latest epoch was requested.
+    *
+    * @param requestedEpoch
+    * @return offset
+    */
+  override def endOffsetFor(requestedEpoch: Int): Long = {
+    inReadLock(lock) {
+      val offset =
+        if (requestedEpoch == latestEpoch) {
+          leo().messageOffset
+        }
+        else {
+          val subsequentEpochs = epochs.filter(e => e.epoch > requestedEpoch)
+          if (subsequentEpochs.isEmpty || requestedEpoch < epochs.head.epoch)
+            UNDEFINED_EPOCH_OFFSET
+          else
+            subsequentEpochs.head.startOffset
+        }
+      info(s"Processed offset for epoch request for epoch:$requestedEpoch and returning offset $offset from epoch list of size ${epochs.size}")
+      offset
+    }
+  }
+
+  /**
+    * Removes all epoch entries from the store with start offsets greater than or equal to the passed offset.
+    *
+    * @param offset
+    */
+  override def clearLatest(offset: Long): Unit = {
+    inWriteLock(lock) {
+      val before = epochs
+      if (offset >= 0 && offset <= latestOffset()) {
+        epochs = epochs.filter(entry => entry.startOffset < offset)
+        flush()
+        info(s"Cleared latest ${before.toSet.filterNot(epochs.toSet)} entries from epoch cache based on passed offset $offset leaving ${epochs.size} in EpochFile for partition $topicPartition")
+      }
+    }
+  }
+
+  /**
+    * Clears old epoch entries. This method searches for the oldest epoch < offset, updates the saved epoch offset to
+    * be offset, then clears any previous epoch entries.
+    *
+    * @param offset the offset to clear up to
+    */
+  override def clearEarliest(offset: Long): Unit = {
+    //TODO - Ben to double check this - committing now as about to get on plane.
+    inWriteLock(lock) {
+      val before = epochs
+      if (offset >= 0 && earliestOffset() <= offset) {
+        val earliest = epochs.filter(entry => entry.startOffset <= offset)
+        if(earliest.size > 0) {
+          epochs = epochs --= earliest
+          //If the offset is less than the earliest offset remaining, add previous epoch back, but with an updated offset
+          if(offset < earliestOffset() || epochs.isEmpty)
```

Comment:
hmm. I think this actually correct. I would like a better way to express it but I can't see one. I'd encourage you to look at the tests in LeaderEpochFileCacheTest to see if you disagree with any of them.